lambda/matching.ml

```diff
@@ -2056,6 +2059,141 @@ let inline_lazy_force arg pos loc =
+                      ap_inlined = Never_inlined;
+                      ap_specialised = Default_specialise;
+                      ap_probe=None
+                    },
+                  (* ... arg *)
+                  varg, Lambda.layout_lazy_contents), Lambda.layout_lazy_contents) ) )
+
+let inline_lazy_force_switch arg pos loc =
+  let idarg = Ident.create_local "lzarg" in
+  let varg = Lvar idarg in
+  let force_fun = Lazy.force code_force_lazy_block in
+  Llet
+    ( Strict,
+      Lambda.layout_lazy,
+      idarg,
+      arg,
+      Lifthenelse
+        ( Lprim (Pisint { variant_only = false }, [ varg ], loc),
+          varg,
+          Lswitch
+            ( varg,
+              { sw_numconsts = 0;
+                sw_consts = [];
+                sw_numblocks = 256;
+                (* PR#6033 - tag ranges from 0 to 255 *)
+                sw_blocks =
+                  [ ( Obj.forward_tag,
+                      Lprim (lazy_forward_field, [ varg ], loc) );
+                    ( Obj.lazy_tag,
+                      Lapply
+                        { ap_tailcall = Default_tailcall;
+                          ap_loc = loc;
+                          ap_func = force_fun;
+                          ap_args = [ varg ];
+                          ap_result_layout = Lambda.layout_lazy_contents;
+                          ap_region_close = pos;
+                          ap_mode = alloc_heap;
+                          ap_inlined = Default_inlined;
+                          ap_specialised = Default_specialise;
+                          ap_probe=None;
+                        } )
+                  ];
+                sw_failaction = Some varg
+              },
+              loc, Lambda.layout_lazy_contents), Lambda.layout_lazy_contents) )
+
+let inline_lazy_force arg pos loc =
+  if !Clflags.afl_instrument then
+    (* Disable inlining optimisation if AFL instrumentation active,
+       so that the GC forwarding optimisation is not visible in the
+       instrumentation output.
+       (see https://github.com/stedolan/crowbar/issues/14) *)
+    Lapply
+      { ap_tailcall = Default_tailcall;
+        ap_loc = loc;
+        ap_func = Lazy.force code_force_lazy;
+        ap_args = [ arg ];
+        ap_result_layout = Lambda.layout_lazy_contents;
+        ap_region_close = pos;
+        ap_mode = alloc_heap;
+        ap_inlined = Never_inlined;
+        ap_specialised = Default_specialise;
+        ap_probe=None;
+      }
+  else if !Clflags.native_code && not (Clflags.is_flambda2 ()) then
+    (* CR vlaviron: Find a way for Flambda 2 to avoid both the call to
+       caml_obj_tag and the switch on arbitrary tags *)
+    (* Lswitch generates compact and efficient native code *)
+    inline_lazy_force_switch arg pos loc
+  else
+    (* generating bytecode: Lswitch would generate too many rather big
+         tables (~ 250 elts); conditionals are better *)
+    inline_lazy_force_cond arg pos loc
+
+
+(* End of lazy implementations. *)
 
```

Comment:
There's a reference to `Obj.forcing_tag` in the implementation of `stdlib/lazy.ml:from_val`, should we delete that too?

Reply:
Ah yes, I've fixed that.  I checked lazy.ml and lazy.mli against the flambda-backend versions and they look ok.